pkg/proxy/iptables/proxier.go

```diff
@@ -564,37 +569,34 @@ func (proxier *Proxier) buildEndpointInfoList(endPoints []hostPortInfo, endpoint
 	return filteredEndpoints
 }
 
-// OnEndpointsUpdate takes in a slice of updated endpoints.
-func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
-	start := time.Now()
-	defer func() {
-		glog.V(4).Infof("OnEndpointsUpdate took %v for %d endpoints", time.Since(start), len(allEndpoints))
-	}()
-
-	proxier.mu.Lock()
-	defer proxier.mu.Unlock()
-	proxier.haveReceivedEndpointsUpdate = true
+type healthCheckEndpoints struct {
+	name      types.NamespacedName
+	endpoints []hostPortInfo
+}
 
+func buildEndpointsMap(allEndpoints []api.Endpoints, hostname string, oldEndpointsMap proxyEndpointsMap) (proxyEndpointsMap, []healthCheckEndpoints, map[endpointServicePair]bool) {
 	activeEndpoints := make(map[proxy.ServicePortName]bool) // use a map as a set
 	staleConnections := make(map[endpointServicePair]bool)
 	svcPortToInfoMap := make(map[proxy.ServicePortName][]hostPortInfo)
 
+	newEndpointsMap := make(proxyEndpointsMap)
+	for key, value := range oldEndpointsMap {
+		newEndpointsMap[key] = value
+	}
+
 	// Update endpoints for services.
 	for i := range allEndpoints {
 		svcEndpoints := &allEndpoints[i]
 
 		// We need to build a map of portname -> all ip:ports for that
 		// portname.  Explode Endpoints.Subsets[*] into this structure.
 		portsToEndpoints := map[string][]hostPortInfo{}
-		for i := range svcEndpoints.Subsets {
-			ss := &svcEndpoints.Subsets[i]
-			for i := range ss.Ports {
-				port := &ss.Ports[i]
-				for i := range ss.Addresses {
-					addr := &ss.Addresses[i]
+		for _, ss := range svcEndpoints.Subsets {
```

Comment:
I did the loops with pointers before because Addresses in particular can get large